### PR TITLE
IBX-6634: Fixed relationship comparison

### PR DESCRIPTION
**Pull request by Steveb-p**
| Question                 | Answer                                              |
|--------------------------|-----------------------------------------------------|
| **JIRA issue**           | [IBX-6634](https://issues.ibexa.co/browse/IBX-6634) |
| **Type**                 | bug                             |
| **Target Ibexa version** | `v4.6`              |
| **BC breaks**            | no                                              |

This PR fixes issues that occur when a Criterion is used that represents a relationship condition.

For example, `LoggedAt` criterion which involved sub-tables with "lower than" condition was overwritten with a "equal to", because comparison operator was ignored - only value was used.

Additionally, a few methods are exposed for use in descendants so that common operations when building a query can be performed without repetition.

_EDIT: Additionally, this also apparently fixes an issue that happens when attempting to `countBy` items, while using sub-tables (pre-joined tables are missing)._

#### Checklist:

- [x] Provided PR description.
- [x] Tested the solution manually.
- [ ] Provided automated test coverage.
- [x] Checked that target branch is set correctly (main for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping for example `@ibexa/php-dev` for back-end changes and/or `@ibexa/javascript-dev` for
  front-end changes).

### CODE REVIEW

**Review by konradoboza — APPROVED** [review]

**Review by bogusez — APPROVED** [review]

**Comment by bogusez** [comment]
The changes are required for example to filtering recent activity logs by object class. Now filtering works correct.

**Comment by Steveb-p** [comment]
This fix has a side-effect on one of the pages in Ibexa DXP, and so needs revision.

**Comment by Steveb-p** [comment]
[2ae1a4c](https://github.com/ibexa/core-persistence/pull/11/commits/2ae1a4c5d30c59a656756c30043ad9e0432c3bb9) seems to fix the issue of additional `SELECT` statements being applied to the `countBy` operation.

**Review by kisztof — APPROVED** [review]

**Review by konradoboza — APPROVED** [review]

**Review by ViniTou — APPROVED** [review]

File: src/contracts/Gateway/AbstractDoctrineDatabase.php [hunk 1]
```diff
@@ -125,18 +125,15 @@ public function countAll(): int
...
-        $qb = $this->connection->createQueryBuilder();
+        $qb = $this->createBaseQueryBuilder();
         $this->applyInheritance($qb);
 
         $identifierColumn = $metadata->getIdentifierColumn();
         $tableAlias = $this->getTableAlias();
-        $qb->select($this->connection->getDatabasePlatform()->getCountExpression($tableAlias . '.' . $identifierColumn));
-        $qb->from($metadata->getTableName(), $tableAlias);
+        $platform = $this->connection->getDatabasePlatform();
+        $qb->select($platform->getCountExpression(sprintf('DISTINCT %s.%s', $tableAlias, $identifierColumn)));
```
Comment (alongosz) [inline src/contracts/Gateway/AbstractDoctrineDatabase.php:134]: There's `\Doctrine\DBAL\Query\QueryBuilder::distinct` but maybe not usable here given it's `COUNT(DISTINCT )` not `DISTINCT COUNT()`?

File: tests/integration/Gateway/ExpressionVisitorTest.php [hunk 2]
```diff
@@ -48,16 +48,38 @@ public function testInvalidField(): void
...
+     */
+    public static function provideForTraversingRelationships(): iterable
+    {
+        yield [
+            new Comparison(
+                'relationship_1.relationship_2.relationship_2_foo',
+                'IN',
+                'bar',
+            ),
+            'relationship_2_table_name.relationship_2_foo IN :relationship_2_foo_0',
```
Comment (alongosz) [inline tests/integration/Gateway/ExpressionVisitorTest.php:72]: TBH I don't follow there relationship between what you declare this relationship and what you get from it (pun intended), but I think we've already discussed this ages ago, so I'll just continue scrolling :D

**Review by alongosz — APPROVED** [review]

File: src/contracts/Gateway/AbstractDoctrineDatabase.php (hunk 1 above)
Comment (Steveb-p) [reply src/contracts/Gateway/AbstractDoctrineDatabase.php:134]: Yes, due to the above it is not usable. I have actually checked and we have a similar code in our `ibexa/core`, and this syntax works for our supported databases.

File: tests/integration/Gateway/ExpressionVisitorTest.php (hunk 2 above)
Comment (Steveb-p) [reply tests/integration/Gateway/ExpressionVisitorTest.php:72]: @alongosz could you rephrase? I don't quite understand what part of the above is the issue? Is it that we are declaring relationships between tables?

**Review by tomaszszopinski — APPROVED** [review]
QA approved on IbexaDXP 4.6 commerce.
